File: Leetcode/dynamic_programming/easy/1137_N-th_Tribonacci_Number.py
class Solution(object):
    def tribonacci(self, n):
        """
        :type n: int
        :rtype: int
        """
        if n == 0:
            return 0

        if n == 1 or n == 2:
            return 1

        fib = [None] * (n + 1)

        fib[0] = 0
        fib[1] = fib[2] = 1

        for i in range(3, n + 1):
            fib[i] = fib[i - 3] + fib[i - 2] + fib[i - 1]

        return fib[n]

File: Leetcode/dynamic_programming/easy/1137_N-th_Tribonacci_Number.cpp
#include <cassert>
#include <vector>
using namespace std;
class Solution {
   public:
    int tribonacci(int n) {  // NOLINT
        vector<int> dp(n + 1);
        if (n == 0) {
            return 0;
        }

        if (n == 1 || n == 2) {
            return 1;
        }

        dp[0] = 0;
        dp[1] = dp[2] = 1;

        for (int i = 3; i < n + 1; ++i) {               // NOLINT
            dp[i] = dp[i - 3] + dp[i - 2] + dp[i - 1];  // NOLINT
        }
        return dp[n];
    }
};

int main() {
    Solution solution;
    assert(solution.tribonacci(4) == 4);
    assert(solution.tribonacci(5) == 7);
    return 0;
}